File: src/main/java/ffblockly/meta/FilterListing.java
package ffblockly.meta;

import java.util.List;
import java.util.regex.Pattern;

import static ffblockly.meta.FilterListing.Attr.*;

public interface FilterListing {
    Pattern FILTERS_LIST_FORMAT = Pattern.compile("^\\s+(?<" +
            timelineSupport +
            ">[T.])(?<" +
            sliceThreading +
            ">[S.])+(?<" +
            commandSupport +
            ">[C.])\\s+(?<" +
            filterName +
            ">\\w+)\\s+(?<" +
            inputSignature +
            ">[AVN\\|]+)\\-\\>(?<" +
            returnSignature +
            ">[AVN\\|]+)\\s+(?<" +
            shortDescription +
            ">.*)$");

    Boolean getTimelineSupport();


    Boolean getSliceThreading();


    Boolean getCommandSupport();


    String getFilterName();


    String getShortDescription();


    List<FilterSignature> getInputSignature();


    List<FilterSignature> getReturnSignature();

    List<AvOption> getOptions();

    enum Attr {
        timelineSupport, sliceThreading, commandSupport, filterName, inputSignature, returnSignature, shortDescription
    }

    enum FilterSignature {
        Audio('A'), Video('V'), Bag('N'), SourceSink('|');
        private char c;
        FilterSignature(char c) {
            this.c = c;
        }
        static FilterSignature valueOf(char c) {
            for (FilterSignature filterSignature : values()) {
                if (c == filterSignature.c) {
                    return filterSignature;
                }
            }
            return null;
        }
    }
}

File: src/main/java/ffblockly/meta/FilterListing.kt
package ffblockly.meta

import java.util.regex.Pattern

interface FilterListing {
    val timelineSupport: Boolean?
    val sliceThreading: Boolean?
    val commandSupport: Boolean?
    val filterName: String
    val shortDescription: String?
    val inputSignature: List<FilterSignature>?
    val returnSignature: List<FilterSignature>?
    val options: List<AvOption>?

    enum class Attr {
        timelineSupport, sliceThreading, commandSupport, filterName, inputSignature, returnSignature, shortDescription
    }

    enum class FilterSignature(private val c: Char) {
        Audio('A'), Video('V'), Bag('N'), SourceSink('|');

        companion object {
            fun valueOf(c: Char): FilterSignature? {
                for (filterSignature in values()) {
                    if (c == filterSignature.c) {
                        return filterSignature
                    }
                }
                return null
            }
        }

    }

    companion object {
        @JvmField
        val FILTERS_LIST_FORMAT = Pattern.compile("^\\s+(?<" + Attr.timelineSupport +
                ">[T.])(?<" + Attr.sliceThreading +
                ">[S.])+(?<" + Attr.commandSupport +
                ">[C.])\\s+(?<" + Attr.filterName +
                ">\\w+)\\s+(?<" + Attr.inputSignature +
                ">[AVN\\|]+)\\-\\>(?<" + Attr.returnSignature +
                ">[AVN\\|]+)\\s+(?<" + Attr.shortDescription +
                ">.*)$")
    }
}